[CL-899] Add web endpoint to show/update home page

lilkraftwerk opened the pull request:
Requesting an initial review for the PR as I imagine there will be some changes, I consider this a WIP still but since it's my first dip into the BE pool I wanted to get an early review and commentary. I'll leave comments in the code on things I'm not sure about or questions I have :)

lilkraftwerk committed:
Homepage endpoint WIP

lilkraftwerk committed:
Home_page show request works

lilkraftwerk committed:
Home page endpoint WIP

lilkraftwerk committed:
Home page endpoint WIP

lilkraftwerk committed:
Merge branch 'CL-5_flexible_pages_iteration_1' of github.com:CitizenLabDotCo/citizenlab into CL-899-add-web-endpoint-to-show-update-home-page

lilkraftwerk committed:
Home page controller WIP

nTraum committed:
Accept nested attributes for pinned_admin_publications directly instead of pins

This change makes the API a bit easier to use because the API client does not need to deal with the join model (Pin) anymore. Instead, they can directly specify a list of admin publication ids as shown in the spec, and Rails will handle the rest in the background. Cool.

lilkraftwerk committed:
Remove console log

lilkraftwerk committed:
Clean up and remove unused code

lilkraftwerk commented on back/app/controllers/web_api/v1/home_pages_controller.rb:
```diff
@@ -0,0 +1,58 @@
+      :cta_signed_out_type,
+      :cta_signed_out_url,
+      :header_bg,
+      pinned_admin_publication_ids: []
```
I couldn't figure out how to get this to work with destroying IDs. For that, the convention I've found is normally `[{ publication_id: 1, _destroy: true}]` but I'm not sure how that works when you're just sending an array of IDs here.

lilkraftwerk commented on back/spec/acceptance/home_pages_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+    patch 'web_api/v1/home_page' do
+      with_options scope: :home_page do
+        parameter :banner_enabled, 'if the banner is enabled'
```
I can fill these out later once the rest of the code seems good.

lilkraftwerk commented on back/spec/acceptance/home_pages_spec.rb:
```diff
@@ -0,0 +1,82 @@
+          expect(json_response.dig(:data, :relationships, :pinned_admin_publications, :data).length).to eq(2)
+        end
+      end
+    end
```
There needs to be at least one more test for removing a pin from the page, but see my comment above, not sure how to do that with the nested attributes.

@nTraum as you mentioned I think we also need tests on the edge cases, to make sure trying to add more than 3 pages fails, etc.

nTraum replied:
Took the liberty to add that test in https://github.com/CitizenLabDotCo/citizenlab/pull/2211/commits/013e4cac48de3dc7a068d48f28c7ec6d493fe351:

I've defined a new context block, where we can a small snippet of code before the test executes. In that `before` block we assign a different list of projects to the home_page - and then execute the test. If we still end up with 2 instead of 3 projects, we know the udpate worked.
If we want to be more specific we could also check for the actual IDs of the projects.

lilkraftwerk replied:
If I understand right:

For the first test we're adding the two projects (via admin publication ids) to a HomePage via the request. In the second test we're setting 3 projects in the code before the test, and then we make the request, which overwrites the in-code association with the two IDs we're sending with the request, and so the length should be 2 at the end?

nTraum replied:
That is correct Lucas! :)

nTraum committed:
[CL-899] Add spec for updating pins of a page

lilkraftwerk committed:
Update docs in spec for home_page

lilkraftwerk commented:
@nTraum updated the docs in the spec and left a comment on your new test. I updated the test you made to check for IDs (rather than length) so please check that to see if it's useful, and I added a further test to check that pins can be removed from a homepage.

lilkraftwerk committed:
Add test for removing pins from page, make previous test more specific

lilkraftwerk committed:
Appease rubocop

nTraum reviewed (approved):
This looks good to me. I think @jinjagit renamed some of the attributes in #2204, which we'd need to do here as well if it hasn't happened yet.

The rest looks great! Welcome back to Ruby world. :gem:

lilkraftwerk commented:
Ahh yes, he told me about that but I haven't merged those changes into this branch yet. Will do that now and ping you for a final go-ahead. Thanks for the review and the help!

lilkraftwerk committed:
Merge branch 'CL-5_flexible_pages_iteration_1' of github.com:CitizenLabDotCo/citizenlab into CL-899-add-web-endpoint-to-show-update-home-page

lilkraftwerk committed:
Rename changed properties for HomePage model

lilkraftwerk commented:
Ok, I changed the name of those attributes. The previous attribute names are only present in the migrations now which I didn't touch
